packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/GroundOverlaysController.java

```diff
@@ -0,0 +1,139 @@
+  }
+
+  @VisibleForTesting
+  GroundOverlaysController(
+      MapsCallbackApi flutterApi,
+      AssetManager assetManager,
+      float density,
+      Convert.BitmapDescriptorFactoryWrapper bitmapDescriptorFactoryWrapper) {
+    this.groundOverlayIdToController = new HashMap<>();
+    this.googleMapsGroundOverlayIdToDartGroundOverlayId = new HashMap<>();
+    this.flutterApi = flutterApi;
+    this.assetManager = assetManager;
+    this.density = density;
+    this.bitmapDescriptorFactoryWrapper = bitmapDescriptorFactoryWrapper;
+  }
+
+  void setGoogleMap(GoogleMap googleMap) {
+    this.googleMap = googleMap;
+  }
+
+  void addGroundOverlays(@NonNull List<Messages.PlatformGroundOverlay> groundOverlaysToAdd) {
+    for (Messages.PlatformGroundOverlay groundOverlayToAdd : groundOverlaysToAdd) {
+      addGroundOverlay(groundOverlayToAdd);
+    }
+  }
+
+  void changeGroundOverlays(@NonNull List<Messages.PlatformGroundOverlay> groundOverlaysToChange) {
+    for (Messages.PlatformGroundOverlay groundOverlayToChange : groundOverlaysToChange) {
+      changeGroundOverlay(groundOverlayToChange);
+    }
+  }
+
+  void removeGroundOverlays(@NonNull List<String> groundOverlayIdsToRemove) {
+    for (@NonNull String groundOverlayId : groundOverlayIdsToRemove) {
+      removeGroundOverlay(groundOverlayId);
+    }
+  }
+
+  @Nullable
+  GroundOverlay getGroundOverlay(@NonNull String groundOverlayId) {
+    GroundOverlayController groundOverlayController =
+        groundOverlayIdToController.get(groundOverlayId);
+    if (groundOverlayController == null) {
+      return null;
+    }
+    return groundOverlayController.getGroundOverlay();
+  }
+
+  private void addGroundOverlay(@NonNull Messages.PlatformGroundOverlay platformGroundOverlay) {
+    GroundOverlayBuilder groundOverlayOptionsBuilder = new GroundOverlayBuilder();
+    String groundOverlayId =
+        Convert.interpretGroundOverlayOptions(
+            platformGroundOverlay,
+            groundOverlayOptionsBuilder,
+            assetManager,
+            density,
+            bitmapDescriptorFactoryWrapper);
+    GroundOverlayOptions options = groundOverlayOptionsBuilder.build();
+    final GroundOverlay groundOverlay = googleMap.addGroundOverlay(options);
+    if (groundOverlay != null) {
+      GroundOverlayController groundOverlayController =
+          new GroundOverlayController(groundOverlay, platformGroundOverlay.getBounds() != null);
+      groundOverlayIdToController.put(groundOverlayId, groundOverlayController);
+      googleMapsGroundOverlayIdToDartGroundOverlayId.put(groundOverlay.getId(), groundOverlayId);
+    }
+  }
+
+  private void changeGroundOverlay(@NonNull Messages.PlatformGroundOverlay platformGroundOverlay) {
+    String groundOverlayId = platformGroundOverlay.getGroundOverlayId();
+    GroundOverlayController groundOverlayController =
+        groundOverlayIdToController.get(groundOverlayId);
+    if (groundOverlayController != null) {
```

Comment:
It is an error to call this with something that cant get a groundOverlayController? If so whos error is it (plugin or host app)?

Reply:
Api has updateGroundOverlays method takes GroundOverlayUpdates object as parameter:
https://github.com/flutter/packages/blob/ff7724c18a803542fc709f942c3bf1cecf7e4864/packages/google_maps_flutter/google_maps_flutter_platform_interface/lib/src/types/ground_overlay_updates.dart
This should be handled by the dart code so that changes only contains items that are available on map.
Not sure would it be good to throw error here if change is triggered for groundoverlay that is not added to the map beforehand. It is indeed misuse of the api, and this is tested in integration tests.